### DISCUSSION TITLE
support for 5th (or more in/outputs)

### DISCUSSION THREAD

**Discussion by bkbartk** [arendst/Tasmota#15606] [Q&A] [answered]
Hello,

I'm planning on building a project which requires 5 servo's.
I'm not fully sure how to build one and another yet, but I noticed that "ShutterRelay5 1" results in an error.

> 18:34:17.234 CMD: ShutterRelay5 1
18:34:17.240 RSL: RESULT = {"Command":"Error"}

on an ESP32 device
I noticed before that a lot configurations only support an array of 4. but until now I didn't really need this.
I don't know it this is a hardware limitation or it's not easily possible to create an array which is larger then 4.

But I was actually wondering if there is a reason for this and if there is some way to work around this issue

**Comment by sfromis** [comment, marked as answer]
It would not be an easy change to extend the support beyond 4, as the settings data structure is space-restricted and each byte is carefully managed. Simplest workaround is another ESP module.

**Comment by bkbartk** [comment]
You mean like esphome or esp-link?
Or do you mean another cpu? Because afaik esp32 is the esp board with the most memory.

**Reply by sfromis** [reply]
What I meant is that when you can have 4 shutters with one ESP module running Tasmota, you could duplicate that for 4 more shutters. Yes, that would be another CPU and so on. Simple workaround, and those boards are not expensive.

The issue with expanding the number for one Tasmota instance is not how much ram the chip has, but due to the architecture of Tasmota, which was built for the ESP8266. The settings data structure is locked at a size of 4 KB, containing "everything" persistent, typically settings, but also persistent values like rules and counters, at fixed reserved locations. In principle, that could be changed for ESP32, but it would be more than a small restructuring of Tasmota to expand on this. Another strategy could be to reimplement shutter support for ESP32 to take advantage of the additional capacity, but that would be quite some work.

While I do not try to say "never", my main point is that additional shutters is not just an easy fix.

**Reply by ascillato2** [reply]
He meant that you can use 2 ESP devices. 1 device with 4 servos and a second device with the extra servo.

The other option, that is modifying Tasmota so as to add this feature, is a lot of work. If you want to spend time on that, please do it and whenever you are able provide a Pull Request for review.

**Comment by bkbartk** [comment]
Thanks to you both,
My project also has some mechanical challenges I need to solve.
I will start with this first, and think about which solution to apply for the 5th servo.
One more question, I know how to compile tasmota, and also how to create a Merge request.
What I don't know is how to debug stuff once I have made changes. Except for using logging.
Is there a way to debug tasmota and step through the lines?

**Reply by s-hadinger** [reply]
Most used way is to add detailed logging. There is JTAG support on ESP32, but it requires specific Pins and hardware. I never tried (yet)
